### PR TITLE
Smhp add features in LCS utills

### PR DESCRIPTION
**Pull request by gmgtamz**
3 scripts

ssh_root.sh to add the SSH root capability
slurm_fix.sh to fix the plugstack bug
pam_adopt_cgroup_wheel.sh to add cgroup, adoptpam and wheel groups

none of them is linked into the python caller script (lifecycle_script.py) yet

### CODE REVIEW

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+
+plugstack_conf="${slurm_dir}/etc/plugstack.conf"
+plugstack_dir="${slurm_dir}/etc/plugstack.conf.d"
+pyxis_conf="${plugstack_dir}/pyxis.conf"
+
+smhp_conf="/opt/ml/config/resource_config.json"
+
+pecho(){
+    #pretty echo
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh:16]: Comment right?

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        node_type="controller-machine"
+    else
+        node_type="login-or-compute"
+    fi
+    # we could revert the search --> no
+    # for hip in $(hostname -I) ;do
+    #     node_type="$(cat "${smhp_conf}" | jq -r ".InstanceGroups[] | select(.Instances[0].CustomerIpAddress == \"${hip}\") | .Name")"
+    #     [[ -n $node_type ]] && break
+    # done
+    pecho "node_type=${node_type}" # controller-machine, worker-group-1
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh:40]: Perhaps in the comment add Ex, it reads more naturally `# Ex: controller-machine, worker-group-1`

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+main(){
+    pecho "start ${FUNCNAME}:"
+
+    check_root
+    get_node_type
+
+    if [[ $node_type == "controller-machine" ]] ;then
+        slurm_fix
+        # scontrol reconfigure
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh:63]: remove this?

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh [hunk 4]
```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+# https://slurm.schedmd.com/pam_slurm_adopt.html
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh:2]: Add a short sentence at the beginning to say what it does, people will not click on every link to understand what it does

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/ssh_root.sh [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    # done
+    pecho "node_type=${node_type}" # controller-machine, worker-group-1
+}
+
+ssh_keys(){
+    pecho "start ${FUNCNAME}:"
+    
+    if ! [[ -f "${ssh_dir}/${key_priv}" ]] ;then
+        rm -f "${ssh_dir}/${key_pub}"
+        # rm -f "${ssh_dir}/${key_auth}"
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/ssh_root.sh:51]: Remove

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh [hunk 6]
```diff
@@ -0,0 +1,249 @@
+#!/bin/bash
+# https://slurm.schedmd.com/pam_slurm_adopt.html
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh:2]: copy paste from other files?

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/ssh_root.sh [hunk 7]
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
```
Comment (mhuguesaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/ssh_root.sh:1]: No generation of ssh keys for root.

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+slurm_conf="${slurm_dir}/etc/slurm.conf"
+slurm_conf_accounting="${slurm_dir}/etc/accounting.conf"
+pam_conf="/etc/pam.d/sshd"
+sshd_conf="/etc/ssh/sshd_config"
+plugstack_conf="${slurm_dir}/etc/plugstack.conf"
+plugstack_dir="${slurm_dir}/etc/plugstack.conf.d"
+pyxis_conf="${plugstack_dir}/pyxis.conf"
+
+smhp_conf="/opt/ml/config/resource_config.json"
+# apt_opts='-o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold"'
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh:36]: remove?

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+squeue
+ssh $node uname
+# Access denied by pam_slurm_adopt: you have no active jobs on this node
+# Connection closed by 10.1.64.87 port 22
+sbatch --wrap "sleep 600" -w $node -N 1
+ssh $node uname
+scancel -u ubuntu
+squeue
+
+# ssh ip-10-1-64-87 uname -a ; -u ubuntu ssh ip-10-1-64-87 uname
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh:233]: remove the below

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+
+exit
+
+
+# test
+node="ip-10-1-64-87"
+squeue
+ssh $node uname
+# Access denied by pam_slurm_adopt: you have no active jobs on this node
+# Connection closed by 10.1.64.87 port 22
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh:227]: ?

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    pecho "leave ${FUNCNAME}."
+}
+
+main $@
+
+exit
+
+
+# test
+node="ip-10-1-64-87"
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh:223]: Static...

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+    pecho "start ${FUNCNAME}:"
+    # follow https://github.com/SchedMD/slurm/blob/master/contribs/pam_slurm_adopt/pam_slurm_adopt.c
+    # https://slurm.schedmd.com/pam_slurm_adopt.html
+    pecho "Adding UsePAM to ${sshd_conf}:"
+    grep -Hn "^UsePAM " ${sshd_conf} 
+    sed -i ${sshd_conf} -e "s/^UsePAM .*/UsePAM yes/g"
+    grep -Hn "^UsePAM " ${sshd_conf} 
+
+    # apt update
+    # # -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold"
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh:77]: remove

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh (hunk 6 above)
Comment (perifaws) [reply 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/pam_adopt_cgroup_wheel.sh:2]: No idea what the files does too, add a header

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh [hunk 13]
```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
```
Comment (mhuguesaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix.sh:1]: Please rename this script to an explicit functionality.

**Comment by perifaws** [comment]
Any update on this @gmgtamz ?

File: 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix_plugstackconf.sh [hunk 14]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+
+# BUG
+# Slurm configless bug workaround for pyxis SPANK plugin reference in etc/plugstack.conf # CF internal ticket D148893639 and V1514329557 # schedMD ticket 20903
```
Comment (perifaws) [inline 1.architectures/5.sagemaker-hyperpod/LifecycleScripts/base-config/utils/slurm_fix_plugstackconf.sh:4]: remove references to the tickets

**Pull request closed without merging**